pkg/v1/cli/catalog/catalog.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package catalog
+
+import (
+	"bytes"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	apimachineryjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+
+	cliv1alpha1 "github.com/vmware-tanzu/tanzu-framework/apis/cli/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/cli/common"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/utils"
+)
+
+const (
+	// catalogCacheFileName is the name of the file which holds Catalog cache
+	// TODO: Use the original catalog file instead of using v2 once the feature is enabled by default
+	catalogCacheFileName = "catalog_v2.yaml"
+)
+
+var (
+	// PluginRoot is the plugin root where plugins are installed
+	pluginRoot = common.DefaultPluginRoot
+)
+
+// ContextCatalog denotes a local plugin catalog for a given context or
+// stand-alone.
+type ContextCatalog struct {
+	sharedCatalog *cliv1alpha1.Catalog
+	plugins       cliv1alpha1.PluginAssociation
+}
+
+// NewContextCatalog creates context-aware catalog
+func NewContextCatalog(context string) (*ContextCatalog, error) {
+	sc, err := getCatalogCache()
+	if err != nil {
+		return nil, err
+	}
+
+	var plugins cliv1alpha1.PluginAssociation
+	if context == "" {
+		plugins = sc.StandAlonePlugins
+	} else {
+		var ok bool
+		plugins, ok = sc.ServerPlugins[context]
+		if !ok {
+			plugins = make(cliv1alpha1.PluginAssociation)
+			sc.ServerPlugins[context] = plugins
+		}
+	}
+
+	return &ContextCatalog{
+		sharedCatalog: sc,
+		plugins:       plugins,
+	}, nil
+}
+
+// Upsert inserts/updates the given plugin.
+func (c *ContextCatalog) Upsert(plugin cliv1alpha1.PluginDescriptor) error { //nolint:gocritic
```

Comment:
**nitpick**: I like including reasons with my `//nolint` comments. (E.g., `//nolint:gocritic // reason …`)

I can't actually tell what linter warning you're suppressing here, and I suspect some future readers of the code may not be able to either; if there's a reason, that may be a hint about whether the method has changed enough that the nolint should be removed.

(If you're looking at the code in an IDE, it's always easy to remove the comment and see what happens, but sometimes we read the code from the GitHub UI, and then the reader is left feeling curious. :smile:)